### PR TITLE
Feature/event interceptor

### PR DESCRIPTION
**Pull request by mtrupkin**
Enable common event logging infrastructure and add additional logging data.

### CODE REVIEW

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 1]
```diff
@@ -199,13 +209,23 @@ private void validatePassword (String userPassword, String rtsPassword) {
...
+    final ProctorEventLogger _eventLogger = new ProctorEventLogger();
+    final EventInfo eventInfo = EventInfo.create(LOGOUT.name(), ENTER.name(), EventParser.getEventDataFields(getHttpCurrentContext().getRequest()));
     try {
+
       ProctorUser thisUser = getThisUser ();
+
+      _eventLogger.putField(PROCTOR_ID.name(), thisUser.getId());
+      _eventLogger.putField(SESSION_ID.name(), thisUser.getSessionKey());
+      _eventLogger.putField(BROWSER_ID.name(), thisUser.getBrowserKey());
+      _eventLogger.info(eventInfo);
```
Comment (trapnine) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: Consider moving this ENTER info log to line 214, immediately after creating the logger. All these user details will get logged in the new EXIT log entry you'll add just before returning true below, or the error entry, should an exception happen.

This way we can see how long it took to run getThisUser(), which is probably the slowest part of this method.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java [hunk 2]
```diff
@@ -64,15 +66,17 @@ public void setBeanFactory (BeanFactory beanFactory) throws BeansException {
 
   @ExceptionHandler ({ NoDataException.class })
   @ResponseBody
-  public ReturnStatus handleNoDataException (NoDataException exp, HttpServletResponse response) {
+  public ReturnStatus handleNoDataException (NoDataException exp, HttpServletRequest request, HttpServletResponse response) {
+    LoggingExceptionHandler.handleException(request, exp);
```
Comment (trapnine) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: Do we need to add something like this to Student?

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    LOGOUT,
+    ADD_ASSESSMENT,
+    EXAM_STATUS,
+    PAUSE_EXAM,
+    PING,
+    PRINT,
+    SEGMENT_TRANSITION,
+    START_EXAM,
+    START_SESSION
+  }
```
Comment (trapnine) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: I suggest going through and removing any of these you didn't use. I removed all of mine in student as the automatic logging with pathInfo() naming took care of most of these.

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    STATUS,
+    EXAMS,
+    SESSION_ID,
+    PROCTOR_ID,
+    BROWSER_ID,
+    REASON,
+    REQUEST_KEY,
+    SEGMENT,
+    REQUEST_COUNT,
+  }
```
Comment (trapnine) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: Same with these, remove any you didn't use.

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventParserFactory.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    if (handler != null && handler instanceof HandlerMethod) {
+      final HandlerMethod handlerMethod = (HandlerMethod) handler;
+      if (handlerMethod.getBean() != null && handlerMethod.getBean() instanceof ActiveSessionXHR) {
+        final ActiveSessionXHR activeSessionXHR = (ActiveSessionXHR)handlerMethod.getBean();
+        final ProctorUser proctorUser = activeSessionXHR.getUser();
+        fields.put(PROCTOR_ID, proctorUser.getId());
+        fields.put(SESSION_ID, proctorUser.getSessionKey());
+        fields.put(BROWSER_ID, proctorUser.getBrowserKey());
+      }
```
Comment (trapnine) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventParserFactory.java]: Cool, I like how you grabbed these values. I hope it isn't too expensive.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java (hunk 1 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: I have included your suggestion.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java (hunk 2 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: I think so.  It looks like student has registered its own exception handling through spring.  Lets make sure the error checkpoint gets logged if there is a problem during the event.

File: proctor/pom.xml [hunk 6]
```diff
@@ -18,7 +18,7 @@
 		<tds-dll.version>3.0.0</tds-dll.version>
 		<progman-client.version>3.0.1</progman-client.version>
 		<sb11-mna-client.version>3.0.0</sb11-mna-client.version>
-		<sb11-shared-code.version>3.0.0</sb11-shared-code.version>
+		<sb11-shared-code.version>3.0.1-SNAPSHOT</sb11-shared-code.version>
```
Comment (millerts) [inline proctor/pom.xml]: There should not be any snapshots in poms moving forward.  It looks like one needs to release the shared code first then bump this version.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 7]
```diff
@@ -199,13 +210,25 @@ private void validatePassword (String userPassword, String rtsPassword) {
   }
 
   public boolean doLogout () {
+    final ProctorEventLogger _eventLogger = new ProctorEventLogger();
```
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: Kinda like you did [here](https://github.com/SmarterApp/TDS_Proctor/pull/23/files#diff-7511aade64a7c4f3b6e2d36e5df3c9d2R84)

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java [hunk 8]
```diff
@@ -253,7 +257,7 @@ protected UserCookie getUserInfo () {
     return _userInfo;
   }
 
-  protected ProctorUser getUser () {
+  public ProctorUser getUser () {
```
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: Can you point me to the code that needs to make this public?  Can you point to the code that needs it public?

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java [hunk 9]
```diff
@@ -64,15 +66,17 @@ public void setBeanFactory (BeanFactory beanFactory) throws BeansException {
...
     _logger.error (exp.getMessage ());
     response.setStatus (HttpServletResponse.SC_OK);
     return exp.getReturnStatus ();
   }
 
   @ExceptionHandler ({ ReturnStatusException.class })
   @ResponseBody
-  public ReturnStatus handleReturnStatusException (ReturnStatusException exp, HttpServletResponse response) {
+  public ReturnStatus handleReturnStatusException (ReturnStatusException exp, HttpServletRequest request, HttpServletResponse response) {
+    LoggingExceptionHandler.handleException(request, exp);
```
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: Can we rename this method `logException`?  I know the class name signals it, but when I see handle I think something is being done with the exception rather than just logging it

File: proctor/src/main/java/TDS/Proctor/Web/presentation/backing/UserDetailsBacking.java [hunk 10]
```diff
@@ -121,7 +131,11 @@ public void setClientName (String clientName) {
     this._clientName = clientName;
   }
 
-  public void init() throws Exception{
+  public void init() throws Exception {
+    final ProctorEventLogger _eventLogger = new ProctorEventLogger();
```
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Web/presentation/backing/UserDetailsBacking.java]: Same comment about initializing at every method call rather than this being injected or created during this class instantiation.

File: proctor/src/main/java/TDS/Proctor/Web/presentation/backing/UserDetailsBacking.java [hunk 11]
```diff
@@ -158,8 +173,14 @@ public void init() throws Exception{
           _selectRolePresenter.createAndUpdateProctorIsCurrent (role,proctorUser.getKey (),getClientName (),role.getEffectiveEntity ().getEntityId (),role.getRoleEntityLevel ().name ());
         }
       }
-      
+
+      _eventLogger.putField(PROCTOR_ID.name(), proctorUser.getId());
```
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Web/presentation/backing/UserDetailsBacking.java]: I'm not sure how this is all wired up, but this block looks similar to [this block](https://github.com/SmarterApp/TDS_Proctor/pull/23/files#diff-17cdff9a3a90ed1aa9f1d079b73ee129R220).  Is there a way we could have the eventLogger know how to log certain object types?  Maybe a logObjectHandler type class that can be injected by Student/Proctor that knows the bits to log for exams, testopportunities, users?

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    EXAMS,
+    SESSION_ID,
+    PROCTOR_ID,
+    BROWSER_ID,
+    REASON,
+    SEGMENT,
+    REQUEST_COUNT,
+  }
+
+  public ProctorEventLogger() {
```
Comment (millerts) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: Why is this needed?  If this is called the object mapper will not be the one that is used by the application so json will not be parsed appropriately since none of the extra modules are configured for the mapper.

File: proctor/src/main/resources/root-context.xml [hunk 13]
```diff
@@ -111,6 +112,9 @@
 		<beans:constructor-arg ref="clientRequestFactory"/>
 	</beans:bean>
 
+	<beans:bean id="objectMapper" class="com.fasterxml.jackson.databind.ObjectMapper" />
```
Comment (millerts) [inline proctor/src/main/resources/root-context.xml]: I'm pretty sure we are already configuring an ObjectMapper with all the necessary modules since Proctor communicates with Exam microservice.  Is this needed or can we find a way to have a single ObjectMapper configured within the spring container?

The object mapper I'm referring to is annotated configuration rather than XML.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java (hunk 7 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: The LoginPresenter code is outside the scope of spring.  I get the impression there are two apps in this project, jsf and spring.  If I am wrong and spring could inject into this class, I would make the change you suggest.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 14]
```diff
@@ -199,13 +210,25 @@ private void validatePassword (String userPassword, String rtsPassword) {
...
+
+      _eventLogger.putField(PROCTOR_ID.name(), thisUser.getId());
+      _eventLogger.putField(SESSION_ID.name(), thisUser.getSessionKey());
+      _eventLogger.putField(BROWSER_ID.name(), thisUser.getBrowserKey());
+
       if (thisUser != null && thisUser.getBrowserKey ().compareTo (Constants.UUIDEmpty) != 0)
         _proctorUserService.logout (thisUser.getKey (), thisUser.getBrowserKey ());
       clearOutProctorCookieInformationOrRemove (true);
+
+      _eventLogger.info(eventInfo.withCheckpoint(EXIT.name()));
```
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: In the case of an exception, the exit checkpoint would not be logged.  The end result logged when there is an exception would be an entry checkpoint log and error checkpoint log.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java (hunk 9 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: Agreed that logException is a better name.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java (hunk 8 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/HttpHandlerBase.java]: https://github.com/SmarterApp/TDS_Proctor/blob/feature/event-interceptor/proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventParserFactory.java#L41

This is where the proctor/session/browser ids are pulled out of the end user's cookie.

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java (hunk 12 above)
Comment (mtrupkin) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: A default constructor is needed for spring aop.  There may be an exact way to define this class without ambiguity on how to use it.

File: proctor/src/main/resources/root-context.xml (hunk 13 above)
Comment (mtrupkin) [inline proctor/src/main/resources/root-context.xml]: I'm made the change and it looks good.
